packages/components/src/number-control/test/index.tsx

```diff
@@ -0,0 +1,616 @@
+			// Third call: invalid, unclamped value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 3, '14', false );
+			// Fourth call: valid, clamped value
+			expect( onChangeSpy ).toHaveBeenNthCalledWith( 4, 10, true );
+		} );
+	} );
+
+	describe( 'Validation', () => {
+		it( 'should clamp value within range on ENTER keypress', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } min={ 0 } max={ 10 } /> );
+
+			const input = getInput();
+
+			await user.clear( input );
+			await user.type( input, '-100' );
+			await user.keyboard( '[Enter]' );
+
+			/**
+			 * This is zero because the value has been adjusted to
+			 * respect the min/max range of the input.
+			 */
+			expect( input ).toHaveValue( 0 );
+		} );
+
+		it( 'should clamp value within range on blur', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } min={ 0 } max={ 10 } /> );
+
+			const input = getInput();
+			await user.clear( input );
+			await user.type( input, '41' );
+
+			// Before blurring, the value is still un-clamped
+			expect( input ).toHaveValue( 41 );
+
+			// Blur the input
+			await user.click( document.body );
+
+			// After blur, value is clamped
+			expect( input ).toHaveValue( 10 );
+		} );
+
+		it( 'should parse to number value on ENTER keypress when required', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } required={ true } /> );
+
+			const input = getInput();
+			await user.clear( input );
+			await user.type( input, 'abc' );
+			await user.keyboard( '[Enter]' );
+
+			expect( input ).toHaveValue( 0 );
+		} );
+
+		it( 'should parse to empty string on ENTER keypress when not required', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } required={ false } /> );
+
+			const input = getInput();
+			await user.clear( input );
+			await user.type( input, 'abc' );
+			await user.keyboard( '[Enter]' );
+
+			// React Testing Library associated `null` to empty string types for
+			// numeric `input` elements
+			// (see https://github.com/testing-library/jest-dom/blob/v5.16.5/src/utils.js#L191-L192)
+			expect( input ).toHaveValue( null );
+		} );
+
+		it( 'should accept empty string on ENTER keypress for optional field', async () => {
+			const user = userEvent.setup( {
+				advanceTimers: jest.advanceTimersByTime,
+			} );
+
+			render( <NumberControl value={ 5 } required={ false } /> );
+
+			const input = getInput();
+			await user.clear( input );
+			await user.keyboard( '[Enter]' );
+
+			// React Testing Library associated `null` to empty string types for
+			// numeric `input` elements
+			// (see https://github.com/testing-library/jest-dom/blob/v5.16.5/src/utils.js#L191-L192)
```

Comment:
As part of the refactor, I switched from `expect( input.value ).toBe( ... )` to `expect( input ).toHaveValue( ... )`.

This change has two consequences, caused by [how Testing Library extracts the value of numeric inputs](https://github.com/testing-library/jest-dom/blob/v5.16.5/src/utils.js#L191-L192):

- the expected value is parsed as a number (previously the tests were expecting strings)
- empty strings are converted to `null`

These are effectively only changes to how the assertions are made, but I wanted to clarify them since there's already enough confusion around the type of the `value` prop.

Reply:
How about `.not.toHaveValue()`? If that works, I believe it will be a better representation of what we're asserting here.

Reply:
I tried that out, and realised that unfortunately [`.not.toHaveValue()` passes for all falsey values](https://github.com/testing-library/jest-dom/blob/948d90f32cc79339bdeebea0454599db74c5d071/src/to-have-value.js#L34) — which means that we wouldn't have a way to tell the difference between `0` and `''`